core/akka-runtime/src/main/scala/com/ing/baker/runtime/akka/actor/process_instance/ProcessInstance.scala

```diff
@@ -73,6 +74,56 @@ object ProcessInstance {
           ingredient
     }
   }
+
+  def getOutputEventName(interactionTransition: InteractionTransition,
+                         log: DiagnosticLoggingAdapter): String = {
+    def getOutputEventNameWithRetryStrategyFiltered(fireEvent: EventDescriptor, eventsToFire: Seq[EventDescriptor]): String = {
+      val outputEventsNames: Seq[String] = eventsToFire.map(_.name)
+      if (outputEventsNames.size != 2)
+        throw new FatalInteractionException(s"Delayed transitions must have 2 input ingredients if FireEventAfterFailure strategy is used")
+      outputEventsNames.filter(_ != fireEvent.name).head
+    }
+
+    def getFirstOutputEventName(eventsToFire: Seq[EventDescriptor]): String = {
+      val outputEvents = interactionTransition.eventsToFire
+      if (outputEvents.size != 1)
+        throw new FatalInteractionException(s"Delayed transitions can only have 1 input ingredient")
+      outputEvents.head.name
+    }
+
+    interactionTransition.failureStrategy match {
+      case FireEventAfterFailure(event) =>
+        getOutputEventNameWithRetryStrategyFiltered(event, interactionTransition.eventsToFire)
+      case RetryWithIncrementalBackoff(_, _, _, _, Some(event)) =>
+        getOutputEventNameWithRetryStrategyFiltered(event, interactionTransition.eventsToFire)
+      case RetryWithIncrementalBackoff(_, _, _, _, None) =>
+        getFirstOutputEventName(interactionTransition.eventsToFire)
+      case BlockInteraction =>
+        getFirstOutputEventName(interactionTransition.eventsToFire)
+      case _ =>
+        log.error("Delayed transition firing with unrecognised Failure Strategy")
+        getFirstOutputEventName(interactionTransition.eventsToFire)
+    }
+  }
+
+  def getWaitTimeInMillis(interactionTransition: InteractionTransition, state: RecipeInstanceState): Long = {
+    val input: immutable.Seq[IngredientInstance] = RecipeRuntime.createInteractionInput(interactionTransition, state)
+    if (input.size != 1)
+      throw new FatalInteractionException(s"Delayed transitions can only have 1 input ingredient")
+    val scalaMillis = try {
```

Comment:
Can't we do some form of pattern matching on input.head.value? I understand what this code does, but it is kind of hard to read.

Reply:
Rewrote this to use a match statement instead of the ugly try catch setup.